modules/metrics_prometheus_v1/src/main/scala/com/evolutiongaming/skafka/metrics/KafkaMetricsCollector.scala

```diff
@@ -0,0 +1,146 @@
+package com.evolutiongaming.skafka.metrics
+
+import cats.Monad
+import cats.syntax.all._
+import com.evolutiongaming.catshelper.CatsHelper._
+import com.evolutiongaming.catshelper.ToTry
+import com.evolutiongaming.skafka.ClientMetric
+
+import scala.jdk.CollectionConverters._
+import scala.util.matching.Regex
+import io.prometheus.metrics.model.registry.MultiCollector
+import io.prometheus.metrics.model.snapshots._
+
+import KafkaMetricsCollector._
+
+/** Prometheus collector for Kafka client metrics.
+  *
+  * Metrics from kafka-clients are pull-based (meaning we have to manually call methods that return metrics), this
+  * doesn't allow us to use the default push-based approach of Prometheus with passing an instance of
+  * `PrometheusRegistry` to the code being observed.
+  *
+  * This class is a workaround for this problem. It's a Prometheus collector that evaluates a given
+  * `F[Seq[ClientMetric[F]]]` on each `collect()` call and returns the result as a list of `MetricFamilySamples`.
+  *
+  * Please note it ignores metrics and labels containing some symbols (like '-' and '.'). This is because Prometheus
+  * doesn't allow such symbols in metric names and label names. Also, it ignores metrics without a name or a
+  * description. See [[https://prometheus.io/docs/practices/naming/]] for more details.
+  *
+  * Example:
+  * {{{
+  * val collectorRegistry: PrometheusRegistry = ??? // your Prometheus Java collector registry
+  * val consumer: Consumer[F, K, V]          = ??? // your consumer
+  * val collector: KafkaMetricsCollector[F]  = new KafkaMetricsCollector[F](consumer.clientMetrics)
+  * collectorRegistry.register(collector)
+  * }}}
+  */
+class KafkaMetricsCollector[F[_]: Monad: ToTry](
+  kafkaClientMetrics: F[Seq[ClientMetric[F]]],
+  prefix: Option[String] = None,
+  customLabels: List[(String, String)],
+) extends MultiCollector {
+
+  private val MetricNameRegex: Regex                 = "[a-zA-Z_:][a-zA-Z0-9_:]*".r
+  private val LabelNameRegex: Regex                  = "[a-zA-Z_][a-zA-Z0-9_]*".r
+  private val (customLabelsKeys, customLabelsValues) = customLabels.separate
+
+  override def collect(): MetricSnapshots = {
+    for {
+      metrics      <- kafkaClientMetrics
+      metricsGroups = metrics.groupBy(m => (m.name, m.group)).values.toList
+      result <- metricsGroups
+        .traverse { metricsGroup =>
+          val metric         = metricsGroup.head
```

Comment:
seems redundant. maybe we could do this only once and not per element of `metricsGroups`?

Reply:
I've simply kept the logic from the previous version of this implementation. Yes, it also seems possible to me, since all the metrics are already grouped by name and should therefore contain equal values. Will change